### ISSUE TITLE
Using params within breadcrumb label

### ISSUE THREAD

**Issue by chad-smith** [udayvunnam/xng-breadcrumb#43]
I'm part way through integrating this library into a mature app and am really impressed with how flexible and powerful it is. Great work!

One quirk of our existing routing is that from a list page `/items`, an individual item would be located at `/items/view/X452`.

The path for this route is expressed as:
```js
{
  path: '/items',
  children: [{
    'view/:id'
  }]
}
```

And the auto generated breadcrumb trail is 
> Home / Items / *view/X452*

I can resolve this on per page basis by using `breadcrumbService.set`, but given the data I want is already in the params, it would be nice for the breadcrumb label to be able to interpolate these values.

A more common use-case for this might be:

```js
{
  path: '/orders',
  children: [{
    ':id',
    data: { breadcrumb: 'Viewing order :id` }
  }]
}

Which would generate

```
> Home / Orders / Viewing order 45

I have assumed a simple interpolation syntax in the breadcrumb label above.

I've had a look at the code and would be reasonably confident creating a pull request for this, but I wanted to raise it as a feature request first to gather your feedback on the idea.

Thanks! 🙂

**Comment by udayvunnam** [comment]
@chad-smith Thank you 😊. This is interesting and is definitely useful to have instead of using `BreadcrumbService.set()`

How to do you plan to extract `:id` in `data: { breadcrumb: `Viewing order :id` }`. We don't have the context of params while defining applications Route config.

Are you thinking of sting manipulation by matching the params or Do we have any implicit solution from Angular Router? 

A PR is definitely welcome to improve this library based on the consumer needs 👍

**Comment by chad-smith** [comment]
I was hoping to have access to the `ActivatedRoute` whilst rendering the breadcrumbs

I've been giving it some thought since posting the issue. My original hope was for this to be generic functionality that worked for all breadcrumb nodes, but I believe you might be able to have nested route params with the same id, e.g. `/groups/:id/items:id` - this would require disambiguation in the placeholder syntax... Maybe it could combine with the alias, something like: `Viewing item @item:id`. I fear this starts becoming very esoteric though.

If they the params were all unique I was considering walking up the ActivatedRoute hierarchy looking for matches in params on `/:(\w)+/`.

If my expectation is correct it might only be feasible to use this approach for the final part of the breadcrumb trail. Would you be OK with that compromise?

I'll get my head into the code a little bit and see what's possible.

**Comment by udayvunnam** [comment]
I was thinking about the same issue of ambiguous params if the param names are same. It will great,
if we provide an implementation that is generic. 

Did you check the usage of *xngBreadcrumb directive - https://github.com/udayvunnam/xng-breadcrumb/blob/089acd97d4e48298a3cd1f5d4bf26a3faac2fc17/apps/breadcrumb-demo/src/app/app.component.html#L68

I think you can implement the pattern as you suggested at app level by fully taking control via breadcrumb template. Let me know if you need any additional content passed to *xngBreadcrumb directive in this regard.

**Comment by udayvunnam** [comment]
Though this is entirely possible. One way is to make breadcrumb as callback that receives resolved route param that returns a string including the id.

```
breadcrumb: (id, ...rest){
 return `Viewing order ${id}`
}
```
See if we can achieve it this way, that becomes generic.

**Comment by udayvunnam** [comment]
@chad-smith I am close to having an implementation similar to 
```
breadcrumb: (id, ...rest){
 return `Viewing order ${id}`
}
```

This will be released in next minor version

**Comment by udayvunnam** [comment]
@chad-smith Thanks for the suggestion and making this library solid. [This feature is available with version 6.2.0](https://github.com/udayvunnam/xng-breadcrumb/releases/tag/v6.2.0) 🎉

https://github.com/udayvunnam/xng-breadcrumb#breadcrumb-as-a-function

**Comment by chad-smith** [comment]
Hi Uday,

I've just updated the library and giving it a test. The first thing that I've come across is that the existing instances of `breadcrumb.set()` have stopped working. Debugging into it and seeing what is expected, I have changed the alias node in my routes to include the `@` prefix and this seems to work. (`breadcrumb: { alias: '@id' } }`).

I am assuming this is a bug, if confirmed, rather than being intentional?

**Comment by udayvunnam** [comment]
Yes this is a bug. We don't need user to specify ~`{breadcrumb: { alias: '@id' } }`~
we just need `{breadcrumb: { alias: 'id' } }` with `breadcrumbService.set('@id', 'MapingLabel')`.

This bug slipped since I changed the tests verifying alias to test breadcrumb as a function. Going to add a fix and additional test for this.

Thanks!

**Comment by guoapeng** [comment]
the component is expressive and stand out of similar component libs like ngx-breadcrumbs, ng-dynamic-breadcrumb, ng2-breadcrumb. even I wrote a blog in my country language to recommend the component. and it's worth a recommendation. 

I am facing the same situation chad-smith was facing. I went through the document of the project to  look for a solution to extract param from the route path. by searching with a lot of keywords and finally got here.
if this useful feature is documented on the tutorial, that would be great!
